contracts/AddressProviderNG.vy

```diff
@@ -0,0 +1,290 @@
+# @version 0.3.10
+"""
+@title CurveAddressProvider
+@author Curve.Fi
+@license Copyright (c) Curve.Fi, 2020-2023 - all rights reserved
+@notice An entrypoint contract for Curve's various registries
+"""
+
+event NewEntry:
+    id: indexed(uint256)
+    addr: address
+    description: String[64]
+
+event EntryModified:
+    id: indexed(uint256)
+    version: uint256
+
+event CommitNewAdmin:
+    deadline: indexed(uint256)
+    admin: indexed(address)
+
+event NewAdmin:
+    admin: indexed(address)
+
+
+struct AddressInfo:
+    addr: address
+    description: String[256]
+    tags: DynArray[String[64], 20]
+    version: uint256
+    last_modified: uint256
+
+
+admin: public(address)
+transfer_ownership_deadline: public(uint256)
+future_admin: public(address)
+
+num_entries: public(uint256)
+check_tag_exists: public(HashMap[String[64], bool])
+check_id_exists: public(HashMap[uint256, bool])
+ids: public(DynArray[uint256, 1000])
+tags: public(DynArray[String[64], 1000])
+id_tag_mapping: HashMap[bytes32, bool]
```

Comment:
Is the idea here that an `id` will not correspond to a specific type of registry anymore as it currently is the case in the AddressProvider? And that instead people should search by tags?

Even if that's not the intended usage, there is a real risk that the link between ID <-> registry type becomes looser. I think that's a slippery slope as it:
- Opens the door to having deprecated registries still on the address provider. This would also make it difficult to tell if a registry is deprecated or not.
- For on-chain integrations, the gas cost of finding the right registry increases with the number of tags and ids. If the integrators can't be sure that an ID corresponds to the reference, most up-to-date version of the registry, they have to use the search.
- I'm not sure tags are great for retrievability. There's no standard taxonomy, casing and spelling might be inconsistent and it will require a human reader. For human-readable info, the description and off-chain docs should be enough.

Reply:
You're right in that there's this risk that the link between ID and registry tag becomes looser and it is a slippery slope.

But for now, I see several aggregators who struggle with new factories of stableswap or cryptoswap invariant amm pools. So: if they simply query `AddressProviderNG.get_ids_for_tags(_tag)`, they can get a list of IDs and use that to integrate for off-chain purposes.

`the gas cost of finding the right registry increases with the number of tags and ids`: I can separate the tags from the AddressInfo struct and that should ameliorate this concern.

Initially I have the following tags: `StableSwap`, `CryptoSwap`, `LLAMMA`, `Router`, `Zap`, `Factory`, `Registry`, `Stablecoin` (for ControllerFactory), `Lending` (for lending ControllerFactory).